packet/src/main/java/org/jclouds/packet/features/DeviceApi.java

```diff
@@ -0,0 +1,145 @@
+import org.jclouds.packet.functions.URIParser;
+import org.jclouds.rest.annotations.Fallback;
+import org.jclouds.rest.annotations.MapBinder;
+import org.jclouds.rest.annotations.PayloadParam;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.annotations.ResponseParser;
+import org.jclouds.rest.annotations.Transform;
+import org.jclouds.rest.binders.BindToJsonPayload;
+import org.jclouds.rest.internal.GeneratedHttpRequest;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.inject.TypeLiteral;
+
+@Path("/projects/{projectId}/devices")
+@Consumes(MediaType.APPLICATION_JSON)
+@RequestFilters({ AddXAuthTokenToRequest.class, AddApiVersionToRequest.class} )
+public interface DeviceApi {
+
+    @Named("device:list")
+    @GET
+    @ResponseParser(ParseDevices.class)
+    @Transform(ParseDevices.ToPagedIterable.class)
+    @Fallback(Fallbacks.EmptyPagedIterableOnNotFoundOr404.class)
+    PagedIterable<Device> list();
+
+    @Named("device:list")
+    @GET
+    @ResponseParser(ParseDevices.class)
+    @Fallback(Fallbacks.EmptyIterableWithMarkerOnNotFoundOr404.class)
+    IterableWithMarker<Device> list(ListOptions options);
+
+    final class ParseDevices extends ParseJson<ParseDevices.Devices> {
+        @Inject
+        ParseDevices(Json json) {
+            super(json, TypeLiteral.get(Devices.class));
+        }
+
+        private static class Devices extends PaginatedCollection<Device> {
+            @ConstructorProperties({ "devices", "meta" })
+            public Devices(List<Device> items, Meta meta) {
+                super(items, meta);
+            }
+        }
+
+        private static class ToPagedIterable extends BaseToPagedIterable<Device, ListOptions> {
+
+            @Inject ToPagedIterable(PacketApi api, Function<Href, ListOptions> linkToOptions) {
+                super(api, linkToOptions);
+            }
+
+           @Override
+           protected List<Object> getArgs(GeneratedHttpRequest request) {
+              return request.getCaller().get().getArgs();
+           }
+
+            @Override
+            protected IterableWithMarker<Device> fetchPageUsingOptions(ListOptions options, Optional<Object> arg0) {
+                return api.deviceApi((String) arg0.get()).list(options);
+            }
+        }
+    }
+
+    @Named("device:create")
+    @POST
+    @Produces(MediaType.APPLICATION_JSON)
+    @MapBinder(BindToJsonPayload.class)
+    @ResponseParser(URIParser.class)
+    URI create(@PayloadParam("hostname") String hostname, @PayloadParam("plan") String plan,
+               @PayloadParam("billing_cycle") String billingCycle, @PayloadParam("facility") String facility,
+               @PayloadParam("features") Map<String, String> features, @PayloadParam("operating_system") String operatingSystem,
+               @PayloadParam("locked") boolean locked, @PayloadParam("userdata") String userdata, @PayloadParam("tags") Set<String> tags);
```

Comment:
For methods with such an amount of parameters, consider creating an object instead.